Added libevent-based evhtp C++ web framework to the benchmarks

OvermindDL1 opened the pull request:
This adds the libevent-based [`evhtp`](https://github.com/criticalstack/libevhtp/) web server library for C++.

Tested on the big server (see the local statistics testing issue #101 with specific comment of:
https://github.com/tbrand/which_is_the_fastest/issues/101#issuecomment-378030590

Tested successfully here, and as expected it is faster than everything else:

╰─➤  bin/benchmarker cpp fasthttprouter iron
Last update: 2018-04-02
```
OS: Linux (version: 4.15.0-041500-generic, arch: x86_64)
CPU Cores: 16
threads: 17, requests: 10000
```
Benchmark running ...
Done. <- fasthttprouter
Done. <- iron
Done. <- evhtp

### Ranking (Framework)

1. [evhtp](https://github.com/criticalstack/libevhtp) (cpp)
2. [iron](https://github.com/iron/iron) (rust)
3. [fasthttprouter](https://github.com/buaazp/fasthttprouter) (go)

### Ranking (Language)

1. cpp ([evhtp](https://github.com/criticalstack/libevhtp))
2. rust ([iron](https://github.com/iron/iron))
3. go ([fasthttprouter](https://github.com/buaazp/fasthttprouter))

### All frameworks

| Language (Runtime)        | Framework (Middleware)    |       Max [sec] |       Min [sec] |       Ave [sec] |
|---------------------------|---------------------------|-----------------|-----------------|-----------------|
| go                        | fasthttprouter            |       11.024293 |       10.558062 |       10.852765 |
| rust                      | iron                      |        7.226015 |        6.806786 |        7.000156 |
| cpp                       | evhtp                     |        6.841014 |        6.082085 |        6.530742 |


─➤  ../stats.exs -w 1 -d 2 evhtp iron fasthttprouter
Total Cores: 16
Concurrent Connections: 1000

Processing servers:
* bin/server_cpp_evhtp
* bin/server_rust_iron
* bin/server_go_fasthttprouter

Processing:  bin/server_cpp_evhtp

* Launching server:  bin/server_cpp_evhtp
* * Warming up for 1s 3 times...
* * * Warming up URL:  http://127.0.0.1:3000/
* * * Warming up URL:  http://127.0.0.1:3000/user/692698444251149697343101647
* * * Warming up URL:  http://127.0.0.1:3000/user
* URL:  http://127.0.0.1:3000/
* * Performing test for 2s...
* * Test Complete
* URL:  http://127.0.0.1:3000/user/4711525600736441384961193490
* * Performing test for 2s...
* * Test Complete
* URL:  http://127.0.0.1:3000/user
* * Performing test for 2s...
* * Test Complete
* Shutting down server:  bin/server_cpp_evhtp

Processing:  bin/server_rust_iron

* Launching server:  bin/server_rust_iron
* * Warming up for 1s 3 times...
* * * Warming up URL:  http://127.0.0.1:3000/
* * * Warming up URL:  http://127.0.0.1:3000/user/90752499589849044673116946278
* * * Warming up URL:  http://127.0.0.1:3000/user
* URL:  http://127.0.0.1:3000/
* * Performing test for 2s...
* * Test Complete
* URL:  http://127.0.0.1:3000/user/93233411904156790889972398607
* * Performing test for 2s...
* * Test Complete
* URL:  http://127.0.0.1:3000/user
* * Performing test for 2s...
* * Test Complete
* Shutting down server:  bin/server_rust_iron

Processing:  bin/server_go_fasthttprouter

* Launching server:  bin/server_go_fasthttprouter
* * Warming up for 1s 3 times...
* * * Warming up URL:  http://127.0.0.1:3000/
* * * Warming up URL:  http://127.0.0.1:3000/user/16849070495628428092531654933
* * * Warming up URL:  http://127.0.0.1:3000/user
* URL:  http://127.0.0.1:3000/
* * Performing test for 2s...
* * Test Complete
* URL:  http://127.0.0.1:3000/user/77865497936674084494369719352
* * Performing test for 2s...
* * Test Complete
* URL:  http://127.0.0.1:3000/user
* * Performing test for 2s...
* * Test Complete
* Shutting down server:  bin/server_go_fasthttprouter



|Path|URL|Errors|Total Requests Count|Total Requests/s|Total Requests Throughput|Total Throughput/s|Req/s Avg|Req/s Std
ev|Req/s Max|Req/s +/-|Latency Avg|Latency Stdev|Latency Max|Latency +/-|50%|75%|90%|99%|
|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|
|bin/server_cpp_evhtp|http://127.0.0.1:3000/|0|1418823|689181.94|86.60MB|42.06MB|88.97k|25.60k|132.02k|58.75%|1.88ms|2.
76ms|27.82ms|88.43%|654.00us|2.05ms|5.23ms|13.09ms|
|bin/server_cpp_evhtp|http://127.0.0.1:3000/user|0|1328222|642693.56|81.07MB|39.23MB|83.18k|24.86k|124.49k|60.00%|1.92m
s|2.91ms|209.02ms|88.89%|671.00us|2.25ms|5.18ms|12.82ms|
|bin/server_cpp_evhtp|http://127.0.0.1:3000/user/0|0|1337805|647171.14|118.65MB|57.40MB|83.80k|24.97k|126.88k|60.00%|1.
90ms|3.01ms|216.83ms|89.06%|647.00us|2.31ms|5.19ms|12.49ms|
|bin/server_go_fasthttprouter|http://127.0.0.1:3000/|0|1099314|524534.17|97.50MB|46.52MB|68.90k|13.14k|103.44k|67.50%|1
.90ms|7.28ms|211.82ms|99.32%|1.08ms|1.69ms|3.28ms|7.92ms|
|bin/server_go_fasthttprouter|http://127.0.0.1:3000/user|0|1039579|510525.5|92.20MB|45.28MB|65.06k|15.45k|101.28k|62.50
%|2.02ms|6.80ms|209.48ms|98.89%|1.15ms|1.97ms|3.51ms|9.06ms|
|bin/server_go_fasthttprouter|http://127.0.0.1:3000/user/0|0|974588|471639.8|152.43MB|73.77MB|60.90k|12.50k|94.06k|70.0
0%|1.95ms|3.83ms|205.04ms|95.50%|1.20ms|2.15ms|3.95ms|10.68ms|
|bin/server_rust_iron|http://127.0.0.1:3000/|0|765084|370090.67|54.72MB|26.47MB|64.12k|41.46k|110.50k|50.83%|237.22us|1
72.50us|11.38ms|88.65%|216.00us|289.00us|396.00us|716.00us|
|bin/server_rust_iron|http://127.0.0.1:3000/user|0|734569|360481.49|52.54MB|25.78MB|61.54k|38.06k|112.53k|50.00%|259.97
us|219.12us|17.47ms|94.71%|226.00us|313.00us|422.00us|800.00us|
|bin/server_rust_iron|http://127.0.0.1:3000/user/0|0|681099|328491.4|85.09MB|41.04MB|42.81k|27.80k|104.28k|58.75%|342.9
6us|284.90us|35.56ms|93.70%|336.00us|434.00us|534.00us|0.95ms|


# Rankings

## Ranking by Average Requests per second:
1. 659682 req/sec : bin/server_cpp_evhtp
2. 502233 req/sec : bin/server_go_fasthttprouter
3. 353021 req/sec : bin/server_rust_iron

OvermindDL1 commented:
I really think we need a document of 'what' needs to be installed globally on the system to allow for compiling/running a given server.  In the case of this C++ server it is only a C++ compiler (gcc/clang/recent-VS/etc) and the CMake C++ Build System, these should be pretty standard on about any developers computer.  The build file for this C++ server will automatically acquire its necessary dependencies (evhtp, libevent, etc...) and as such does not require them to be installed system-wide.

OvermindDL1 commented:
On an aside, should I make a PR for my tiny statistical benchmarker?  It's an Elixir script that wraps the `wrk` tester, and as such it does require `wrk` to be installed on the path to be able to run it (as well as elixir of course).

waghanza commented:
Hi @OvermindDL1,

> I really think we need a document of 'what' needs to be installed globally on the system to allow for compiling/running a given server

Sure of that (in my point of view) but `docker` helps doing this :stuck_out_tongue:  (we can use **official** images, so not having to document manually). For me documentation has to come after results stability

> On an aside, should I make a PR for my tiny statistical benchmarker?

I think we should use / contribute to https://github.com/sdogruyol/cryload as mentionned in #172, and why not being inspired by 
+https://github.com/rakyll/hey
+ https://github.com/wg/wrk
+ https://github.com/JoeDog/siege

> If anyone else can do a quick review and test on their own system

My results :

~~~
Last update: 2018-04-03
```
OS: Linux (version: 4.15.12-201.fc26.x86_64, arch: x86_64)
CPU Cores: 8
threads: 9, requests: 10000
```
Benchmark running ...
Done. <- fasthttprouter
Done. <- iron
Done. <- evhtp

### Ranking (Framework)

1. [iron](https://github.com/iron/iron) (rust)
2. [evhtp](https://github.com/criticalstack/libevhtp) (cpp)
3. [fasthttprouter](https://github.com/buaazp/fasthttprouter) (go)

### Ranking (Language)

1. rust ([iron](https://github.com/iron/iron))
2. cpp ([evhtp](https://github.com/criticalstack/libevhtp))
3. go ([fasthttprouter](https://github.com/buaazp/fasthttprouter))

### All frameworks

| Language (Runtime)        | Framework (Middleware)    |       Max [sec] |       Min [sec] |       Ave [sec] |
|---------------------------|---------------------------|-----------------|-----------------|-----------------|
| go                        | fasthttprouter            |        9.722964 |        8.280072 |        8.959467 |
| rust                      | iron                      |        7.951282 |        6.351917 |        7.012645 |
| cpp                       | evhtp                     |        7.491679 |        6.975121 |        7.191633 |
 ~~~

waghanza commented:
> it is not programmed well enough
so we can improve it :stuck_out_tongue_winking_eye:

OvermindDL1 commented:
> it is not programmed well enough
> so we can improve it stuck_out_tongue_winking_eye

Except it's Crystal.  I've not yet seen crystal perform as well as go/rust/c++ on many-core machines.  Single-core it's fine, but on many-core it fails pretty hard.  Just as an example, run right now cryload was able to perform 32250 requests a second, where wrk reached 662037 requests a second to the same evhtp server:

```zsh
╰─➤  ./cryload -s http://localhost:3000/ -n 100000
Preparing to make it CRY for 100000 requests!
Total request made: 10000
Total request made: 20000
Total request made: 30000
Total request made: 40000
Total request made: 50000
Total request made: 60000
Total request made: 70000
Total request made: 80000
Total request made: 90000
Completed All Requests!
-------------------------------

Time taken per request:
Min: 0.021 ms
Max: 1.332 ms
Average: 0.032 ms

Requests Statistics:
Request p/s: 31092.40039547703
2xx requests 0
Non 2xx requests 100000
Total request made: 100000
Total time taken: 3.2162200000018935 seconds

╰─➤  wrk -t 16 -c 1000 -d 3 http://localhost:3000/
Running 3s test @ http://localhost:3000/
  16 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.98ms    4.14ms 218.16ms   92.09%
    Req/Sec    46.16k    18.15k  114.09k    74.18%
  2247243 requests in 3.10s, 137.16MB read
Requests/sec: 724878.98
Transfer/sec:     44.24MB
```

`wrk` just absolutely *blows* it away.  I'm thinking that Crystal's (as a language) thread handling is broken in it's core currently, and I'm running this version of it:

```zsh
─➤  crystal --version
Crystal 0.24.1 (2017-12-22)

LLVM: 4.0.0
Default target: x86_64-unknown-linux-gnu
```

OvermindDL1 commented:
As a comparison, here's how Crystal's router_cr tests:

```zsh
╰─➤  ./cryload -s http://localhost:3000/ -n 100000                                                                 1 ↵
Preparing to make it CRY for 100000 requests!
Total request made: 10000
Total request made: 20000
Total request made: 30000
Total request made: 40000
Total request made: 50000
Total request made: 60000
Total request made: 70000
Total request made: 80000
Total request made: 90000
Completed All Requests!
-------------------------------

Time taken per request:
Min: 0.019 ms
Max: 1.369 ms
Average: 0.029 ms

Requests Statistics:
Request p/s: 33995.82123366732
2xx requests 0
Non 2xx requests 100000
Total request made: 100000
Total time taken: 2.941537999998844 seconds


╰─➤  wrk -t 16 -c 1000 -d 3 --latency http://localhost:3000/
Running 3s test @ http://localhost:3000/
  16 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    12.29ms   27.22ms 423.30ms   98.23%
    Req/Sec     5.95k     2.19k   21.12k    77.23%
  Latency Distribution
     50%    9.85ms
     75%   10.29ms
     90%   14.90ms
     99%  150.49ms
  266094 requests in 3.10s, 15.73MB read
Requests/sec:  85839.18
Transfer/sec:      5.08MB
```

See, cryload is showing the same results as the C++ server, yet wrk shows that the crystal server is *SUBSTANTIALLY* slower, in fact it looks like almost the perfect speed comparison of 1/16th (16 cores) minus an overhead percentage, which implies to me that either Crystal is not multi-core (or has something monumentally stupid like a GIL or so), or is just so inefficient so as to mimic it being single-core slow.

OvermindDL1 commented:
Thus using cryload would be entirely useless as it cannot saturate a server, just like the current crystal client cannot saturate a server.  (Why is an untested/new language like *crystal* of all things being used to run the actual benchmark siege?!)

OvermindDL1 commented:
Also looking at cryload's code I cannot find any statistical testing at all, the most it reports is simple time comparisons like at https://github.com/sdogruyol/cryload/blob/master/src/cryload/request.cr#L17 but without jitter testing, error bounds, outliers, etc... thus it is not accurate in that way either, all of which are things that `wrk` does, and that is in addition to it not able to actual load down a server.

Hmm, let's test CPU usage to the Crystal server:
* `cryload` pegs out at 100% CPU usage the entire test, the Crystal router_cr server averages ~46% CPU usage, even though it is using ~16 threads!
* `wrk` ~134% average CPU load (so 1.3 CPU cores worth), the Crystal router_cr server averages 100% CPU usage, even though it is using ~16 threads!

And the C++ server:
* `cryload` pegs out at 100% CPU usage (so a single core worth again), the C++ evhtp server averages ~54% CPU usage.
* `wrk` averages ~596% CPU usage (almost 6 cores worth!), the C++ evhtp server averaged ~964% CPU Usage, so almost *10* cores worth!

In essence, cryload cannot saturate any server, it cannot even saturate a crystal server, so I don't know how it is supposed to make a server 'cry' to use their own terminology.  ;-)

Where `wrk` hammers the crystal router_cr into a pegged single-core, even though the crystal router_cr is using 15/16 threads, thus implying Crystal has a GIL or something, and of course both `wrk` and the C++ evhtp server absolutely peg the entire system down to every scrap of CPU power possibly available.

Let's check memory usage too, going to just use `wrk` for this since the crystal benchmarker cannot even saturate it's own crystal servers:
* Crystal router_cr:  ~800megs
* C++ evhtp:  < 160megs

So why on *earth* is crystal using so much memory when it is so inefficient as well?  The C++ evhtp server probably used a lot less than 160megs but the way I was checking the check became inaccurate at < 0.1% total installed memory so I just rounded up.  And of course I could link in a better memory allocator if I wanted the C++ server to be even faster too (same memory allocator Rust uses actually) though it would use ~30% more memory then.

OvermindDL1 commented:
Testing the other linked ones.

# hey

### Crystal router_cr:

```zsh
─➤  ~/go/bin/hey -c 1000 -z 30s http://localhost:3000/

Summary:
  Total:        30.0219 secs
  Slowest:      0.1684 secs
  Fastest:      0.0001 secs
  Average:      0.0300 secs
  Requests/sec: 46528.0889
  

Response time histogram:
  0.000 [1]     |
  0.017 [121224]        |∎∎∎∎∎∎
  0.034 [867444]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.051 [10045] |
  0.067 [276]   |
  0.084 [10]    |
  0.101 [12]    |
  0.118 [3]     |
  0.135 [37]    |
  0.152 [339]   |
  0.168 [609]   |

Latency distribution:
  10% in 0.0165 secs
  25% in 0.0176 secs
  50% in 0.0221 secs
  75% in 0.0242 secs
  90% in 0.0263 secs
  95% in 0.0285 secs
  99% in 0.0343 secs

Details (average, fastest, slowest):
  DNS+dialup:   0.0001 secs, 0.0001 secs, 0.1684 secs
  DNS-lookup:   0.0000 secs, 0.0000 secs, 0.0006 secs
  req write:    0.0000 secs, 0.0000 secs, 0.0092 secs
  resp wait:    0.0271 secs, 0.0000 secs, 0.0361 secs
  resp read:    0.0015 secs, 0.0000 secs, 0.1312 secs

Status code distribution:
  [200] 1396863 responses
```

`hey` used ~366% CPU usage, router_cr used 100% again.

### C++ evhtp

```zsh
╰─➤  ~/go/bin/hey -c 1000 -z 30s http://localhost:3000/

Summary:
  Total:        30.0193 secs
  Slowest:      0.2621 secs
  Fastest:      0.0000 secs
  Average:      0.0300 secs
  Requests/sec: 44407.2818
  

Response time histogram:
  0.000 [1]     |
  0.026 [465463]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.052 [522155]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.079 [10976] |∎
  0.105 [391]   |
  0.131 [15]    |
  0.157 [39]    |
  0.183 [417]   |
  0.210 [88]    |
  0.236 [423]   |
  0.262 [32]    |

Latency distribution:
  10% in 0.0001 secs
  25% in 0.0146 secs
  50% in 0.0290 secs
  75% in 0.0314 secs
  90% in 0.0355 secs
  95% in 0.0460 secs
  99% in 0.0593 secs

Details (average, fastest, slowest):
  DNS+dialup:   0.0001 secs, 0.0000 secs, 0.2621 secs
  DNS-lookup:   0.0000 secs, 0.0000 secs, 0.1580 secs
  req write:    0.0000 secs, 0.0000 secs, 0.1187 secs
  resp wait:    0.0001 secs, 0.0000 secs, 0.0798 secs
  resp read:    0.0153 secs, 0.0000 secs, 0.1551 secs

Status code distribution:
  [200] 1333077 responses
```

`hey` used ~480% CPU usage, evhtp used ~150%...

## Conclusions

Thus `hey` uses a lot more CPU to hit a *lot* less load than `wrk`, in addition it couldn't even saturate the C++ server, that is abysmally bad as a tester when it cannot even saturate what it's trying to test...  It's looking like `hey` is only doing IO work in a single thread but replicating processing out to multiple threads.  This explains why it cannot saturate the C++ server while capping out at only around ~480% CPU usage, *AND* it matches that Go's libevent interfaces default to a single-thread processor, this is not a fault of Go but rather a fault of `hey`.  Even if that single-threaded processor issue was fixed `hey` I don't see getting anywhere near what `wrk` can do due to it's processing and GC overhead (as shown by how much CPU it was using just to calculate a single thread's worth of IO...).



# Siege

Ah siege, an oldie that I used to use before `wrk`, reason why switched is embedded through-out the rest of this section, but I am nostalgic to it.  ^.^

### Crystal router_cr:

```zsh
─➤  siege -c 1000 -b -t 30s http://localhost:3000/
** SIEGE 3.0.8
** Preparing 1000 concurrent users for battle.
The server is now under siege...[error] socket: read error Connection reset by peer sock.c:479: Connection reset by peer

Lifting the server siege...      done.

Transactions:                 610771 hits
Availability:                 100.00 %
Elapsed time:                  29.24 secs
Data transferred:               0.00 MB
Response time:                  0.05 secs
Transaction rate:           20888.20 trans/sec
Throughput:                     0.00 MB/sec
Concurrency:                  967.96
Successful transactions:      610771
Failed transactions:               1
Longest transaction:            7.58
Shortest transaction:           0.00
```

`siege` used ~476% CPU usage, router_cr used 120% (yes more!  This is because of a kernel context switching cost that siege induces in servers, and thus makes it pretty poor for benchmarking).

### C++ evhtp

```zsh
╰─➤  siege -c 1000 -b -t 30s http://localhost:3000/
** SIEGE 3.0.8
** Preparing 1000 concurrent users for battle.
The server is now under siege...
Lifting the server siege...      done.

Transactions:                1863797 hits
Availability:                 100.00 %
Elapsed time:                  29.45 secs
Data transferred:               0.00 MB
Response time:                  0.02 secs
Transaction rate:           63286.82 trans/sec
Throughput:                     0.00 MB/sec
Concurrency:                  977.79
Successful transactions:     1863797
Failed transactions:               0
Longest transaction:            7.20
Shortest transaction:           0.00
```

`siege` used ~114% CPU usage, evhtp used ~406%, yes siege is this inefficient, the other reason I switched from it long ago...

## Conclusions

So yes, you can see why I switched from siege to `wrk` many years ago.  In addition it is a bit faulty in it's reporting because it doesn't report header sizes and so forth in its calculations.

# Final Conclusion

I've yet to find or see anything anywhere that beats `wrk` in raw throughput, reporting accuracy, and efficiency.  Though I would *love* to be shown one that actually does.  :-)

So no, cryload, hey, and siege are not good options.

OvermindDL1 commented:
Did some testing with the crystal benchmarker built in to this repo, it cannot even get close to saturating the C++/Rust/Go servers and it's capping out at ~100% CPU Usage as well, which also implies it has a GIL and/or is otherwise useless at benchmarking, so the built in client is useless.

* client ~100% CPU Usage : cpp_evht ~103% CPU Usage (better than even cryload!  o.O)
* client ~93% CPU Usage : crystal_router_cr ~100% (The server got capped out so the client couldn't even hit 'its' own max speed...)
* client ~113% CPU Usage : rust_iron ~118%
* client ~119% CPU Usage : go_fasthttprouter ~115%

And the results were:

> > ╰─➤  bin/benchmarker cpp router_cr iron fasthttprouter                                                         > [8/1850]
> Last update: 2018-04-03
> ```
> OS: Linux (version: 4.15.0-041500-generic, arch: x86_64)
> CPU Cores: 16
> threads: 17, requests: 10000
> ```
> Benchmark running ...
> Done. <- router_cr
> Done. <- fasthttprouter
> Done. <- iron
> Done. <- evhtp
> 
> ### Ranking (Framework)
> 
> 1. [iron](https://github.com/iron/iron) (rust)
> 2. [evhtp](https://github.com/criticalstack/libevhtp) (cpp)
> 3. [router_cr](https://github.com/tbrand/router.cr) (crystal)
> 4. [fasthttprouter](https://github.com/buaazp/fasthttprouter) (go)
> 
> ### Ranking (Language)
> 
> 1. rust ([iron](https://github.com/iron/iron))
> 2. cpp ([evhtp](https://github.com/criticalstack/libevhtp))
> 3. crystal ([router_cr](https://github.com/tbrand/router.cr))
> 4. go ([fasthttprouter](https://github.com/buaazp/fasthttprouter))
> 
> ### All frameworks
> 
> | Language (Runtime)        | Framework (Middleware)    |       Max [sec] |       Min [sec] |       Ave [sec] |
> |---------------------------|---------------------------|-----------------|-----------------|-----------------|
> | crystal                   | router_cr                 |        9.271305 |        7.377060 |        8.109800 |
> | go                        | fasthttprouter            |       11.308970 |       10.808372 |       11.063619 |
> | rust                      | iron                      |        7.676771 |        5.964357 |        6.809752 |
> | cpp                       | evhtp                     |        7.443327 |        6.498350 |        6.914891 |


So this definitely shows that the built-in client is pretty abysmal as well.  If anything it's measuring the CPU Kernel Context Switching costs more than it is benchmarking the servers...  It cannot saturate more than a single core, and thus is artificially limiting *most* servers, this is pretty ***bad*** and makes the results reported in the Readme entirely false...  Was this never checked for before?!  o.O
